packages/protocol/contracts/common/LibraryProxy.sol

```diff
@@ -0,0 +1,86 @@
+pragma solidity ^0.5.13;
+/* solhint-disable no-inline-assembly, no-complex-fallback, avoid-low-level-calls */
+
+import "openzeppelin-solidity/contracts/utils/Address.sol";
+import "./Registry.sol";
+
+/**
+ * @title A library proxy utilizing the Unstructured Storage pattern, allowing
+ * for calls to a linked library pointed to by a Registry.
+ * @dev This contract should be inherited from (e.g. for a library called Math,
+ * create `contract MathProxy is LibraryProxy`). The inheriting contract should
+ * implement a fallback function that calls `_delegateToRegisteredLibrary`,
+ * passing in a Registry hashed identifier referencing the implementation
+ * library.
+ */
+contract LibraryProxy {
+  // Used to store the address of the registry contract.
+  // NOTE: non-standard EIP1967 value to not clash with the calling contract's
+  // underlying Proxy.
+  bytes32 private constant REGISTRY_POSITION = bytes32(
+    uint256(keccak256("eip1967.proxy.registry")) - 1
+  );
+
+  function _setRegistry(address registryAddress) public {
+    bytes32 registryPosition = REGISTRY_POSITION;
+
+    require(Address.isContract(registryAddress), "Invalid contract address");
+
+    // Store the address of the implementation contract in an explicit storage slot.
+    assembly {
+      sstore(registryPosition, registryAddress)
+    }
+  }
+
+  function _getRegistry() public view returns (address registry) {
+    bytes32 registryPosition = REGISTRY_POSITION;
+    // Load the address of the implementation contract from an explicit storage slot.
+    assembly {
+      registry := sload(registryPosition)
+    }
+  }
+
+  function _delegateToRegisteredLibrary(bytes32 libraryIdentifier) internal {
+    Registry registry = Registry(_getRegistry());
+
+    address implementationAddress = registry.getAddressForOrDie(libraryIdentifier);
+
+    // Avoid checking if address is a contract or executing delegated call when
+    // implementation address is 0x0
+    require(implementationAddress != address(0), "No Implementation set");
+    require(Address.isContract(implementationAddress), "Invalid contract address");
```

Comment:
is this check necessary? it protects only against malicious registry behavior

Reply:
True. We have the same sort of check in the regular Proxy, which also only really sanity checks Governance behavior. Is this check particularly costly?

Reply:
Maybe not so costly as an individual tx, but I'd say that since it guards the gateway to ALL libraries calls, it will get called A LOT, and that will add up.

Reply:
The `extcode` family of opcodes requires state access. So it is fairly high on the cost scale

Reply:
🌈